### ISSUE TITLE
Performance opportunity for yarn 3 cache

### ISSUE THREAD

**Issue by belgattitude** [actions/setup-node#325]
Thanks for the new cache feature. Much easier.

After few weeks, I realized that while it supports yarn, there's some improvements that can be made.

Yarn 3 (probably yarn 2+ too) manages downloaded archives pretty well (`.yarn/cache/*.zip`) and invalidating on yarn.lock changes does not take that into account and I saw a lot of cache misses.

As an example I converted back to action-cache to illustrate and test.

The P/R is here https://github.com/belgattitude/nextjs-monorepo-example/pull/369.

I'm wondering if a similar approach could be done with setup-node ?

PS: Key points

<details>
  <summary>Example with action cache</summary>

```
      - name: Get yarn cache directory path
        id: yarn-cache-dir-path
        run: echo "::set-output name=dir::$(yarn config get cacheFolder)"
      - name: Restore yarn cache
        uses: actions/cache@v2
        id: yarn-cache 
        with:
          path: ${{ steps.yarn-cache-dir-path.outputs.dir }}       
          key: yarn-cache-folder-os-${{ runner.os }}-node-${{ env.node-version }}-${{ hashFiles('**/yarn.lock', '.yarnrc.yml') }}      
          restore-keys: |
            yarn-cache-folder-os-${{ runner.os }}-node-${{ env.node-version }}-
            yarn-cache-folder-os-${{ runner.os }}-
```
</details>

<details>
  <summary>Result after a new package was added</summary>

```
yarn install --immutable
➤ YN0000: ┌ Fetch step
  ➤ YN0013: │ 1719 packages were already cached, one had to be fetched (superjson@npm:1.7.5)
➤ YN0000: └ Completed in 1s 124ms

Cache Size: ~127 MB (133261279 B)
Cache saved successfully
Cache saved with key: yarn-cache-folder-os-Linux-node--f118ea4bee07eada9df36ad2e83fd6febcbaf06b5b8962689c7650659e872ad3
```
</details>

I

**Comment by merceyz** [comment]
It can be even more efficient by not including the OS and Node version in the cache key as well - https://github.com/actions/setup-node/pull/272#issuecomment-873564091

**Comment by nickmccurdy** [comment]
This only applies to zero install, right? In that case there probably needs to be a way to detect or toggle it.

**Comment by dmitry-shibanov** [comment]
Hello @belgattitude. Thank you for your feature request. Could you please describe it little bit. Do you want to add restore-keys to the action to take previous cache or you also want to add `.yarn/cache/*.zip` to cached directories ?

**Comment by dmitry-shibanov** [comment]
Hello @belgattitude, just a gentle ping.

**Comment by belgattitude** [comment]
@dmitry-shibanov first of all thanks for considering this issue. I've edited the P/R desc with latest comments from @merceyz 

> Do you want to add restore-keys to the action to take previous cache

Yes that's the idea, that way we take advantage of the built-in yarn 2+/3+ cache management. Yarn will prune old packages and only fetch new ones keeping the packages zipped into `cacheFolder` in sync with package.json (works with workspace/monorepo too).

Note that if multiple actions are run in parallel, only the first one will be able to reserve the cache, others will print a warning after:

```
> Post restore yarn cache
   Unable to reserve cache with key yarn-cache-folder- 
   cb79cb8e7c0fdd859237ac516da53bc2fdafceb3230fd68444aa4f55c9238980, another job may be creating this cache.
```

As far as I tested it does not create problems and can be safely ignored..


> you also want to add .yarn/cache/*.zip to cached directories

Actually this what matters most. AFAIK setup/node try to cache the `node_modules` folder(s). 

I would add this should be the only thing that needs to be cached (but there I'm not 100% sure).

There's some kind of guarantee that yarn cache is portable across os and node versions as @merceyz pointed out, while node_modules might not (native binaries, postinstall tricks). 

As the cache folder is configurable (through [yarnrc.yml](https://yarnpkg.com/configuration/yarnrc#cacheFolder) or `YARN_CACHE_FOLDER` env), a good way to read it

```yaml
      # Get the yarn cache path.
      - name: Get yarn cache directory path
        id: yarn-cache-dir-path
        run: echo "::set-output name=dir::$(yarn config get cacheFolder)"
```

The restore key could be a constant:

```yaml
      - name: Restore yarn cache
        uses: actions/cache@v2
        id: yarn-cache # use this to check for `cache-hit` (`steps.yarn-cache.outputs.cache-hit != 'true'`)
        with:
          path: ${{ steps.yarn-cache-dir-path.outputs.dir }}
          key: yarn-cache-folder-${{ hashFiles('**/yarn.lock', '.yarnrc.yml') }}
          restore-keys: |
            yarn-cache-folder-
```

Interestingly pnpm has a [benchmark](https://pnpm.io/benchmarks) page with some insights about the different modes (cache+node_modules, cache only)

![image](https://user-images.githubusercontent.com/259798/143036812-efed0f26-8100-49a6-8b61-3db73ae2473e.png)

Looks having both node_modules + yarn cache will be faster (excluding action/cache restore/generate...). I guess even more as yarn won't probably have to re-run the link phase (generally slow with native binaries: esbuild, swc, sharp, prisma...). But with caching node_modules I'm not sure of the portability though (it depends also of setup/node)

I would say a safe way would be to just save 'yarn cache folder' and not node_modules.

Let me know if it answer your questions.

Have a great day


PS: @merceyz thanks for your comment about portability, I was wondering if the newly introduced [supportedtArchitecture](https://yarnpkg.com/configuration/yarnrc#supportedArchitectures) change something ?

**Comment by devgioele** [comment]
Do you recommend using `actions/setup-node` to cache `node_modules` and to use `actions/cache` to cache "the yarn cache"?
Caching both would allow us to skip the linking step, right? I don't see why it cannot be done.

**Comment by mklueh** [comment]
Deleting yarn.lock and running yarn install helped, but in my case with yarn caching enabled

**Comment by nickmccurdy** [comment]
I think this is a duplicate of #328, which is similar but not specific to yarn

**Comment by belgattitude** [comment]
Just to share, here's my updated composite action for yarn 3+ with node-modules linker.

https://gist.github.com/belgattitude/042f9caf10d029badbde6cf9d43e400a

**Comment by merceyz** [comment]
> It can be even more efficient by not including the OS and Node version in the cache key as well - [#272 (comment)](https://github.com/actions/setup-node/pull/272#issuecomment-873564091)

This got a bit more complicated in v3.1.0 with the introduction of [`supportedArchitectures`](https://yarnpkg.com/configuration/yarnrc#supportedArchitectures), if `current` is included in any of the options (the default) then the cache changes based on the setup.

**Comment by nijikon** [comment]
Is there any plan to add this anytime soon?

**Comment by marko-zivic-93** [comment]
Hello @nijikon,
We have great news for you. We will be taking this feature and working on it in the upcoming quarter, so you can expect this to be available soon... 😃

**Comment by dsame** [comment]
Hello @belgattitude, please confirm the problem we want to save is eliminating the whole cache on any lock file invalidating.

I mean that if only one (of 2) dependency changed https://github.com/akv-demo/setup-node-test/commit/44f68b499db0435dd6e4ca019d88ac4a017129d1#diff-51e4f558fae534656963876761c95b83b6ef5da5103c4adef6768219ed76c2deL19

the whole cache is reset https://github.com/akv-demo/setup-node-test/actions/runs/4767849996/jobs/8476525793#step:3:82

and all (of 2) dependencies must be fetched https://github.com/akv-demo/setup-node-test/actions/runs/4767849996/jobs/8476525793#step:4:28

Is your suggestion to solve the described problem with `restore-keys`-like behaviour?

**Comment by belgattitude** [comment]
Hey @dsame 

>  please confirm the problem we want to solve is eliminating the whole cache reset on any lock file invalidating.

Yes.

But https://github.com/akv-demo/setup-node-test/actions/runs/4767849996/jobs/8476525793#step:4:28 does not seem to work ? Am I wrong ?

**Comment by belgattitude** [comment]
BTW you can have a look to some ideas

- https://github.com/strapi/strapi/pull/16581 (I've commented in the code see the changed files)
- https://github.com/belgattitude/compare-package-managers/tree/main/.github/actions
- https://gist.github.com/belgattitude/042f9caf10d029badbde6cf9d43e400a

**Comment by dsame** [comment]
> Hey @dsame
> 
> > please confirm the problem we want to solve is eliminating the whole cache reset on any lock file invalidating.
> 
> Yes.
> 
> But https://github.com/akv-demo/setup-node-test/actions/runs/4767849996/jobs/8476525793#step:4:28 does not seem to work ? Am I wrong ?

Yes, it is `a proof the problem exists` build - pay attention it uses v3(i.e. current) action, this way i confirmed we really need the feature.

**Comment by dsame** [comment]
@belgattitude with the close look i started to doubt about the feature.

First of all yarn has 2 modes - local and global cache depending on `enableGlobalCache` option in `.yarnrc.yml`

With the global cache yarn does not purge the obsolete dependencies and we can not restore the previous cache in order to avoid the endless grow

proof: you [can see different versions of lodash](https://github.com/akv-demo/setup-node-test/actions/runs/4821512135/jobs/8587481166#step:12:42)

with the local cache yarn [does purge](https://github.com/akv-demo/setup-node-test/actions/runs/4821512135/jobs/8587481166#step:7:29) the obsolete dependencies BUT, in this case you should not use actions cache at all - you should keep `.yarn` in your repo instead as [it is recommended in the yarn docs]

(https://yarnpkg.com/features/offline-cache). 

If we even apply the feature only for local caches we are getting into the danger of clashing `.yarn` from cache with `.yarn` from repo. It is still possible to workaround this but it looks to be useful only for marginal case: local cache + .yarn not included to the repo

Can you please provide more justification to use caching instead of adding `.yarn` to the repo?

Also i have a question about node modules? Do we really need it to cache it with yarn? Since PnP has been introduced `yarn` does not have it at all, but have `.pnp.cjs` instead, which is assumed to be added to the repo as well.

**Comment by belgattitude** [comment]
From my phone, so writing is minimal. 

Override global cache with YARN_ENABLE_GLOBAL_CACHE=false. That gives you an uniform way to do it for ci. The yarn cache will be stored in .yarn/cache by default but you can get it from config as well. 

Just cache .yarn/cache not .yarn

Better to avoid caching node modules with this approach. It generally does not bring speed up due to action/cache extra work. That also solve the pnp support.

**Comment by dsame** [comment]
Hello @belgattitude - one more clarification: is there any reason to do not add `.yarn/cache` to the repository as it is [recommended in the yarn docs](https://yarnpkg.com/features/offline-cache) and how the yarn team do for [their own repo](https://github.com/yarnpkg/berry/tree/master/.yarn/cache) ?

Having `yarn/cache` saves the Git storage and gives better performance than caching, but i suspect there may be some specific reason to use cache instead of adding to the repo?

**Comment by belgattitude** [comment]
The initial offline mode idea isn't probably used. Storing large binaries in git and their mutation is probably not a good idea. I guess after few updates you'll end up with a git repo of 10's of Gb. 
Git isn't a cache for large binaries. Rewritting history to purge obsolete entries isn't easy.

That said I can see why they explored this. But the default is to not commit the cache. 

But it's true setup/node action cache feature shouldn't be enabled by the user if he's using full offline.

**Comment by dsame** [comment]
Thank you @belgattitude for your opinion, now i agree it is worth to continue with the feature in order to avoid huge repo. 

Can you please take a look at this Proof of concept draft PR and tell you opinion about does it make a sense to include node version and step id into the cache key keeping in mind having many huge cache will hit storage limit soon

https://github.com/actions/setup-node/pull/744/files#diff-55f15e2366942ad15f71a41ac983f8ce9a9882b28b7fd9082f3a26c799783064R42

**Comment by dsame** [comment]
@belgattitude can you please also confirm the following :

We have to use 3rd parameter of restoreCache in case if all of 3 conditions met:


The 3rd parameter of `restoreCache` has to be set to not null in case if all of 3 conditions met:

 1. package manager is yarn
 1. enableGlobalCache is omitted or false
 1. there's no .yarn/cache directory in the working dir

**Comment by belgattitude** [comment]
@dsame 

1.  only from v2
2. (mmmh). on CI this should be overriden to enableGlobalCache: false. Don't run with true on CI (see previous comment)
3. yes

I have looked into your PR. stepId not necessary in my opinion. But there's something missing to ensure yarn 4. Not a lot of time the next days... I'll figure out a way to share info asap

**Comment by belgattitude** [comment]
In the meantime I've try to add a lot of useful info in

- https://github.com/strapi/strapi/pull/16581
- https://github.com/strapi/strapi/pull/16638

Might be worth to have a quick glance

**Comment by dsame** [comment]
The next iteration of [PoC PR](https://github.com/actions/setup-node/pull/744/files#diff-55f15e2366942ad15f71a41ac983f8ce9a9882b28b7fd9082f3a26c799783064R48) showed the detection of [all the conditions](https://github.com/actions/setup-node/issues/325#issuecomment-1536649126) implies the significant changes in code base and makes it unreasonable complicated. 

@belgattitude can you please take a look at this [issue](https://github.com/actions/setup-go/issues/358) - does not it seem the goal of reusing the existing cache can be achieved with resolving that issue in the more straightforward and simpler way? 

I looked through the links you've sent, 
 - according to the key-prefix idea, it is clear but you've provided a custom solution for the specific build while we are trying to provide universal solution for the whole action - this is why it become complicated
 - according to including `.yarnec.yaml` in hash calculation, it is the overkill and cause unnecessary cache refreshes
 - according to node_modules caching it brings performance gain only for cases of some binaries have to be rebuilt from sources and it is not considered now to be effective enough for most users - downloading and unpacking lot of files can be event slower than `npm install ...` if the dependencies are already cached.
 
 Summary: can you please confirm or deny that `key-prefix` input added to the `setup-node` cache can boost the yarn3 performance and achieve the same goal as auto-generating the same key-prefix if [the conditions are met](https://github.com/actions/setup-node/issues/325#issuecomment-1536649126)?

**Comment by belgattitude** [comment]
Sorry I couldn't read properly but some quick toughts:

> an you please take a look at this https://github.com/actions/setup-go/issues/358 - does not it seem the goal of reusing the existing cache can be achieved with resolving that issue in the more straightforward and common way?

I would agree that params helps.

> ccording to including .yarnec.yaml in hash calculation, it is the overkill and cause unnecessary cache refreshes

yarnrc.yml isn't something that is mutated often, so I wouldn't worry about it to invalidate things often. 

Not accounting for it looks fragile in my impression, but I'm not 100% sure. I would let the benefit to doubt and include it in the hash (ie someone changes an advanced param...)

What do you mean by overkill ? is the hash call slow ?

> according to node_modules caching it brings performance gain only for cases of some binaries have to be rebuilt from sources and it is not considered now to be effective enough for most users - downloading and unpacking lot of files can be event slower than npm install ... if the dependencies are already cached

For yarn 2+, I would not cache node_modules at all. (only few edges case would benefit from it, in this case they might use a totally different action than the one in setup/node)

> Can you please confirm or deny that key-prefix input added to the setup-node cache can boost the yarn3 performance and achieve the same goal as auto-generating the same key-prefix if https://github.com/actions/setup-node/issues/325#issuecomment-1536649126?

The general idea seems good, but as always a deep testing is required. Do you expect me to do it ? In that case I'll need some time.

> showed the detection of https://github.com/actions/setup-node/issues/325#issuecomment-1536649126 implies the significant changes in code base and makes it unreasonable complicated.

Can you precise the meaning of ? "Unreasonable" =>  "in your opinion you wouldnt include the feature  ?". Or you're still wondering ? I would understand of course. 

But as a small thing, it might have an impact on global CI time... I feel it might worth it.

**Comment by nijikon** [comment]
@marko-zivic-93 back in April https://github.com/actions/setup-node/issues/325#issuecomment-1499149677, you wrote that you will be having a look at this in the upcoming quarter. Were you talking about Q2 or Q3 of 2023?

**Comment by dsame** [comment]
Hello @nijikon the requested changes  [are merged](https://github.com/actions/setup-node/commit/e33196f7422957bea03ed53f6fbb155025ffc7b8) into the main branch and are available since [v3.7.0 release](https://github.com/actions/setup-node/releases/tag/v3.7.0)

**Comment by nijikon** [comment]
@dsame thanks. I will have a look if that helps my case.

**Comment by dsame** [comment]
Hello @nijikon, i am going to close this issue because the PRs are merged and due to inactivity, but please feel free to reopen it or create new issue in case if the problem still exists

**Comment by nijikon** [comment]
@dsame I think I'm good. My current problem is that it does not cache the `node_modules` folder, but I read that it's expected.
